lti_consumer/lti_1p3/extensions/rest_framework/authentication.py

```diff
@@ -0,0 +1,75 @@
+"""
+Django REST Framework extensions for LTI 1.3 & LTI Advantage implementation.
+
+Implements a custom authentication class to be used by LTI Advantage extensions.
+"""
+from django.utils.translation import ugettext as _
+from rest_framework import authentication
+from rest_framework import exceptions
+
+from lti_consumer.models import LtiConfiguration
+
+
+class Lti1p3ApiAuthentication(authentication.BaseAuthentication):
+    """
+    LTI 1.3 Token based authentication.
+
+    Clients should authenticate by passing the token key in the "Authorization".
+    LTI 1.3 expects a token like the following:
+        Authorization: Bearer jwt-token
+
+    Since the base implementation of this library uses JWT tokens, we expect
+    a RSA256 signed token that contains the allowed scopes.
+    """
+    keyword = 'Bearer'
+
+    def authenticate(self, request):
+        """
+        Authenticate an LTI 1.3 Tool.
+
+        This doesn't return a user, but let's the external access and commit
+        changes.
+        TODO: Consider creating an user for LTI operations, both to keep track
+        of changes and to use Django's authorization flow.
+        """
+        auth = request.headers.get('Authorization', '').split()
+        lti_config_id = request.parser_context['kwargs'].get('lti_config_id')
+
+        # Check if auth token is present on request and is correctly formatted.
+        if not auth or auth[0].lower() != self.keyword.lower():
+            msg = _('Missing LTI 1.3 authentication token.')
+            raise exceptions.AuthenticationFailed(msg)
+
+        if len(auth) == 1:
+            msg = _('Invalid token header. No credentials provided.')
+            raise exceptions.AuthenticationFailed(msg)
+
+        if len(auth) > 2:
+            msg = _('Invalid token header. Token string should not contain spaces.')
+            raise exceptions.AuthenticationFailed(msg)
+
+        # Retrieve LTI configuration or fail if it doesn't exist
+        try:
+            lti_configuration = LtiConfiguration.objects.get(pk=lti_config_id)
+            lti_consumer = lti_configuration.get_lti_consumer()
+        except:
+            msg = _('LTI configuration not found.')
+            raise exceptions.AuthenticationFailed(msg)
+
+        # Verify token validity
+        # This doesn't validate specific permissions, just checks if the token
+        # is valid or not.
+        try:
+            lti_consumer.check_token(auth[1])
+        except:
+            msg = _('Invalid token signature.')
+            raise exceptions.AuthenticationFailed(msg)
+
+        # Passing parameters back to the view through the request.
+        # Not exactly optimal.
+        request.lti_configuration = lti_configuration
+        request.lti_consumer = lti_consumer
+
+        # Return (None, None) since this isn't tied to any authentication
+        # backend on Django, and it's just used for LTI endpoints.
+        return (None, None)
```

Comment:
Do you think an LTI user can be used to pass the LTI configuration back to the view?

Reply:
We can do that, but it'll add 2 complexities to this task:

1. We'll have to create a user (in playbooks/configuration) and add setting to specify which is the LTI user.
2. We'll need to create a model somewhere to link a token to a LTI session, and pass back the correct LTI consumer to the client.

Given the complexities, I've opted to avoid touching Django authentication backends and passing back a user.
From the Django POV, this this is a public viewset. This design choice will also removes the worry of accidentally authenticating a user and give him permissions to other APIs, and effectively isolates LTI auth for others views in the system.

Are you ok if I keep the current implementation as it? 
Is the thing I've described above unnecessarily complex?

Reply:
@viadanna I've added a comment in the code above explaining the reason I went with this approach.

Reply:
Yeah, too much complexity to make it worthwhile 👍